### ISSUE TITLE
Adopt "clear linux" patches?

### ISSUE THREAD

**Issue by colemickens** [NixOS/nixpkgs#63708]
Has anyone spent much time looking at Clear Linux and the patches they carry that get them such noticeable perf bumps?

I'd be curious to see if we could (optionally) support some of those patches in nixpkgs?

**Comment by jcumming** [comment]
There is a lot of valuable technology in Clear Linux (like the auto proxy stuff) that I'd love to see in NixOS. 

One of the reasons Clear Linux is so fast is that the Clear Linux team at Intel does *tons* of profiling and optimization.

**Comment by matthewbauer** [comment]
I think having a NixOS profile or module to do this would be very useful! it would be interesting to see how close to Clear Linux performance we can get.

/cc @volth

I don't think it's a good idea to make this the default, however. There is always a good reason these things are not made the default for GCC, Linux, and other software. For instance, Clear Linux only supports Westmere and later Intel CPUs, as well as disabling. Based on the Phoronix article (https://www.phoronix.com/scan.php?page=article&item=ubuntu-clear-tweaks&num=1) and https://github.com/InBetweenNames/gentooLTO, here are some of the flags that could be useful:

## GCC
arch: westmere
tune: haswell
configure flags:

```
--enable-offload-targets=nvptx-none
--disable-libunwind-exceptions
--enable-lto
```

(possible) C flags:

```
-O3
-feliminate-unused-debug-types
--param=ssp-buffer-size=32
-Wl--copy-dt-needed-entries-m64-fasynchronous-unwind-tables
-fasynchronous-unwind-tables
-fno-semantic-interposition
-ffat-lto-objects
-fno-signed-zeros
-fno-trapping-math
-fassociative-math
-fexceptions
-ftree-loop-distribute-patterns
-Wl,-sort-common
-fno-semantic-interposition
-fipa-pta
-fdevirtualize-at-ltrans
```

## Glibc

```--enable-tunables=...```

## Linux

patches: https://github.com/clearlinux-pkgs/linux

## Systemd

patches: https://github.com/clearlinux-pkgs/systemd

**Comment by volth** [comment]
> I it would be interesting to see how close to Clear Linux performance we can get.

Do we have some sort of independent benchmark suite?

> For instance, Clear Linux only supports Westmere and later Intel CPUs, as well as disabling. 

I see a problem here: many virtual machines (including NixOS own test-driver as https://github.com/NixOS/nixpkgs/pull/59855 is not merged yet) run without exposing CPU type to the guest, so the software compiled with ```-march=westmere``` would likely crash with "invalid instruction" being run in VM, even if the physical CPU is ```westmere``` or newer.

**Comment by davidak** [comment]
>Do we have some sort of independent benchmark suite?

@volth i packaged phoronix-test-suite recently. https://github.com/NixOS/nixpkgs/pull/62509

**Comment by vcunat** [comment]
I suppose the discussion should be more about the individual changes, probably in separate threads/tickets.  Generally though, if some patches/settings are so good, why don't they get integrated/defaulted upstream?

#### GCC:

- IIRC system-wide `-O3` is not recommended, as it generally allows expensive trade-offs that you probably only want to take on performance-critical parts (and still maybe not all of them)
- `arch`, `tune`: these are about improving performance on some CPUs at the expense of making others slower or not working at all &ndash; that doesn't seem suitable for general-purpose distribution/binaries (general-purpose in terms of CPU choice).  Adding yet another "platform" (in terms of rebuilding and distributing) would be quite expensive, so the best I expect is for interested people to build everything themselves, but then they might prefer even closer tailoring, e.g. `-march=native`.
- LTO: opensuse now [enables this system-wide](https://hubicka.blogspot.com/2019/05/gcc-9-link-time-and-inter-procedural.html) and only disables it on particular packages, so that sounds OK, but we should first upgrade the default compiler to 8 or 9

**Comment by OpenSourceAnarchist** [comment]
@vcunat I agree that creating multiple platforms for binary redistribution would be inefficient and having users build packages themselves is a much better solution. As a previous GentooLTO user, I would LOVE to have an easy way to build everything with `-march=native -O3 (or -O2 -ftree-vectorize) LTO, and graphite optimizations`. Then, being able to override individual packages with "toned down" flags if necessary. This is a very Gentoo-y thing and there isn't a reason to _not_ allow NixOS users to do the same thing if desired, so long as they accept the consequences and additional work involved.

I would also love to see LTO enabled tree-wide after a GCC/LLVM upgrade. Theoretically, every package _should_ work with it, so finding packages and reporting bugs upstream would be extremely beneficial.

I'm being completely honest here: NixOS could capture a huge percentage of Gentoo users by allowing users to have these features (setting custom CFLAGS/CXXFLAGS/LDFLAGS and overriding for packages). There is a whole group of people who will do anything for an extra 1% performance gain, even if it is a placebo effect!

**Comment by vcunat** [comment]
Well, [-march](https://github.com/NixOS/nixpkgs/pull/59225) is now configurable (in a pure way, of course), so you might push towards adding more such options.  I don't think it's difficult to add.  But beware, nix will force way more recompilation on you than gentoo did.  (Per-package CFLAGS/... overrides seem rather easy already.)

**Comment by volth** [comment]
As a quick hack which would allow to make some experiments, one could supply many flags separated by space in ```gcc.tune``` (```gcc.tune = "skylake -O3"````)
```gcc.arch```is checked against the list of known architectures, but ```gcc.tune``` is just an opaque string to pass down to gcc flags

**Comment by OpenSourceAnarchist** [comment]
@volth That is quite a dirty way of doing it :) I might try a test config with that later...

@vcunat I'm talking about something like [this](https://github.com/InBetweenNames/gentooLTO/blob/master/sys-config/ltoize/files/package.cflags/ltoworkarounds.conf) specifically, and in general, all of [these](https://github.com/InBetweenNames/gentooLTO/tree/master/sys-config/ltoize/files) files/workarounds used by GentooLTO. Having overlays like this becomes extremely difficult to maintain in Nix (basically a separate overlay for every package with a different stdenv with workarounds for each of these).

If you're up for a radical suggestion, allow users to specify a global (and per-package if desired) list of C/CXX/LDFLAGS along with successive lists of FLAGS if the previous one fails. This is something that GentooLTO does not do that would make such a project _so much easier_ to declare in one file.

Essentially, **if** `-march=skylake -O3 -flto` fails to compile/pass tests, **then** try again with `-march=skylake -O3 -flto -ffat-lto-objects`, **then** `-march=skylake -O3 -fno-lto`, **then** `-march=skylake -O2 -fno-lto` and so on. Literally solves 95% of the issues with local builds and aggressive optimizations. The rest can be written per-package (though some packages have their own optimizations in the build phase which could cause trouble with this method... just a suggestion!).

**Comment by vcunat** [comment]
You don't really need to override stdenv for individual packages; you can e.g. override those packages directly to amend `CFLAGS` or `NIX_CFLAGS*` passed to `stdenv.mkDerivation`.

**Comment by OpenSourceAnarchist** [comment]
How does that work? The only documentation I found was on a similar thread on Discourse about overlays for `-march=native` on specific packages (https://discourse.nixos.org/t/targeting-particular-x86-64-generation-haswell-ivybridge-skylake/2280/24).

**Comment by devhell** [comment]
Can't wait for a "ricer" option. :laughing:

**Comment by OpenSourceAnarchist** [comment]
@devhell You jest, but I am very eager to switch from a riced GentooLTO install to NixOS when it becomes easy to build all locally optimized, aggressively compiled packages :)

**Comment by davidak** [comment]
@OpenSourceAnarchist that sounds interesting. are there any studies or benchmarks how much you can improve the performance with that on a recent CPU? Can you estimate how many % it can bring?

**Comment by OpenSourceAnarchist** [comment]
@davidak It is definitely an interesting project that, on average, tends to be roughly 10% faster overall (compared to standard -O2 compilation). In select workloads, an optimized setup will yield a nearly 200% faster package on modern CPUs compared to -O2. Of course, compiling from source takes some time, though an entire world rebuild on my laptop (Dell Latitude 5300) easily finishes overnight.

Stats can be seen (with a grain of salt since dependent libraries aren't compiled with these flags) here: https://www.phoronix.com/scan.php?page=article&item=gcc9-skylake-avx512

Some packages work OOTB with aggressive optimizations, others need workarounds (see GentooLTO's work here: https://github.com/InBetweenNames/gentooLTO/blob/master/sys-config/ltoize/files/package.cflags/ltoworkarounds.conf). All in all, I'd love to see NixOS support more aggressive optimizations (with workarounds added as-needed) for those who are willing to take the time and risk of using them. Obviously it would have to be an exclusively opt-in feature since it would most likely violate purity and could break after any given update. Though the atomic feature of NixOS is what makes me want to switch to it from Gentoo. If a package update breaks on Gentoo, it's difficult to find out what compiler flag caused an issue. On NixOS, I can rollback and test different combinations until everything works. Absolutely ingenious system that any source-based distro should adopt.

**Comment by matthewbauer** [comment]
> Some packages work OOTB with aggressive optimizations, others need workarounds (see GentooLTO's work here: https://github.com/InBetweenNames/gentooLTO/blob/master/sys-config/ltoize/files/package.cflags/ltoworkarounds.conf). All in all, I'd love to see NixOS support more aggressive optimizations (with workarounds added as-needed) for those who are willing to take the time and risk of using them. Obviously it would have to be an exclusively opt-in feature since it would most likely violate purity and could break after any given update.

Those workarounds are pretty extensive though! I think we can accomplish those with overlays, although it's a little scary how much can break from LTO. Maybe an external project could try those out.

**Comment by OpenSourceAnarchist** [comment]
It's unfortunate that LTO (and even -O3) breaks so many packages. It _shouldn't_ by design, but projects like these help upstream find bugs so less workarounds are needed in the future.

I'll be willing to test and contribute bug reports to any "NixLTO" efforts :laughing:. I'm still learning Nix and NixOS in general, but if this gets any traction I'll switch over to NixOS full-time and contribute workarounds through overlays that I find. Definitely could convince some GentooLTO folks to try out NixOS too for this project.

In fact, testing packages for build failures would be way easier with NixOS compared to Gentoo!

**Comment by mdedetrich** [comment]
>
It's unfortunate that LTO (and even -O3) breaks so many packages. It shouldn't by design, but projects like these help upstream find bugs so less workarounds are needed in the future.

Does it break so many packages though? Clear Linux is already having a large package repository and it doesn't appear to be causing that many issues.

I agree though that the sooner we find the breakages with LTO/O3 the sooner they can be fixed. Nixos already has great infrastructure in place to support patching of packages anyways.

**Comment by vcunat** [comment]
No, LTO should break only a relatively small fraction of packages, and it's an option (in principle) suitable for enabling by default.  I already posted [this SUSE link](https://hubicka.blogspot.com/2019/05/gcc-9-link-time-and-inter-procedural.html); I expect we'd be able to piggy-back on their work on fixing some of the breakages, too.

**Comment by Atemu** [comment]
Still important to me.

**Comment by Atemu** [comment]
Crazy idea: Periodically assimilate the clear-linux optimisation declarations and add an attr to mkDerivation that lets you delare a clear-linux package name to automatically apply the optimisations (clflags, patches etc.) of.

**Comment by Lillecarl** [comment]
https://xanmod.org/ Here's a kernel currently being built for NixOS with some clearlinux patches and other goodies. I intend to test Anbox with this kernel soon-ish

**Comment by WillPower3309** [comment]
> https://xanmod.org/ Here's a kernel currently being built for NixOS with some clearlinux patches and other goodies. I intend to test Anbox with this kernel soon-ish
> 
> `boot.kernelPackages = with pkgs.linuxKernel.packages; linux_xanmod;`

Small nitpick but the expression can be slightly simplified to `boot.kernelPackages = pkgs.linuxPackages_xanmod;`
I've been using it with great success! :)

**Comment by Lillecarl** [comment]
> > https://xanmod.org/ Here's a kernel currently being built for NixOS with some clearlinux patches and other goodies. I intend to test Anbox with this kernel soon-ish
> > `boot.kernelPackages = with pkgs.linuxKernel.packages; linux_xanmod;`
> 
> Small nitpick but the expression can be slightly simplified to `boot.kernelPackages = pkgs.linuxPackages_xanmod;` I've been using it with great success! :) 

Sure, but then you can't find it by searching on https://search.nixos.org which is something i value higher than character count.

**Comment by WillPower3309** [comment]
Does Intel provide a way for users to see what compilation flags they use on the clear Linux packages? Phoronix benchmarks always list them yet I haven't been able to find them myself. I'm thinking if they can be verified via benchmarking to provide a significant performance increase, I'd make some PRs to nixpkgs implementating the more performant flags.

**Comment by Atemu** [comment]
You can simply have a look at their sources; they've got packaging scripts.

**Comment by GrabbenD** [comment]
Any updates to this?
I'm new to NixOS and I'm unsure how to achieve this in a clean way

**Comment by gytis-ivaskevicius** [comment]
low hanging fruit to increase performance would be to provide targets for `x86_64-v3`/`x86_64-v4` (not related to clear linux patches, only related to improved performance)

**Comment by SuperSamus** [comment]
#202526

**Comment by ShalokShalom** [comment]
Why dont we just outright build the Clear Linux kernel?
Arch is doing that, and I raised an issue above this very comment. 
Feel free to comment there. 🙂

**Comment by GrabbenD** [comment]
> Why dont we just outright build the Clear Linux kernel?

Xanmod kernel has patches from Clear Linux and it's available as a NixOS package.

**However, most of Clear Linux's performance isn't coming from the kernel but rather how the packages are compiled and various performance patches: https://github.com/clearlinux-pkgs/linux**

**Comment by ShalokShalom** [comment]
I understand. [ALHP](https://somegit.dev/ALHP/ALHP.GO) provides an alternative repo, and includes LTO, -03 and v3. Would that be an option?

**Comment by GrabbenD** [comment]
ALHP repository isn't compatible with NixOS if that's the question

But if you're wondering if ALHP is an alternative to CachyOS then yes. You can actually use CachyOS repositories in a existing Arch Linux installation (check their wiki) and it's even possible to use both ALHP and CachyOS simultaneously (by placing one above the other to select priority in /etc/pacman.conf and leaving Arch Linux repositories at the bottom as fallback). I personally prefer CachyOS as it offers LTO, O3, v4/v3/v2 and I found it to be more stable / less prone to breakage (but that's off-topic so let's not discuss it further 🙂).

I [hacked together a OSTree](https://github.com/GrabbenD/ostree-utility) setup using Dockerfiles meaning I can have a imperative system, rollbacks AND optimized packages :heart:

**Comment by ShalokShalom** [comment]
I was wondering if the _approach_ - providing an entire repository - would work for us.

**Comment by Atemu** [comment]
@ShalokShalom we don't need entire new repos or infrastructure, we already have all of that. The real meat of this is getting individual patches and tweaks into individual packages. With Nix we can hide all of them by default behind a opt-in override flag anyways (or perhaps even a `config` attribute).

**Comment by ShalokShalom** [comment]
So what is missing? Can you link me to the according issue or PR?

**Comment by mikunimaru** [comment]
There is a lack of documentation on how to add build flags to most packages (including the kernel). Without this, other users will not be able to experiment with the build.

**Comment by 06kellyjac** [comment]
https://nixos.wiki/wiki/Build_flags

https://nixos.wiki/wiki/Linux_kernel

https://nixos.org/guides/nix-pills/nixpkgs-overriding-packages.html

https://ryantm.github.io/nixpkgs/using/overrides/

**Comment by WillPower3309** [comment]
> So what is missing? Can you link me to the according issue or PR?

There's nothing missing to get this implemented, nixpkgs already supports adding patches and defining flags. Unless we want an issue tracking every individual package it's just up to the community to (optionally) add those patches and flags to packages that interest them.

**Comment by ShalokShalom** [comment]
> There is a lack of documentation on how to add build flags to most packages (including the kernel). Without this, other users will not be able to experiment with the build.

I dont understand. Why not add complete packages with the new binaries?

**Comment by WillPower3309** [comment]
> > There is a lack of documentation on how to add build flags to most packages (including the kernel). Without this, other users will not be able to experiment with the build.
> 
> I dont understand. Why not add complete packages with the new binaries? 
> 

Because nix is a source based Linux package manager. Build flags are added as default values in a package definition that can be optionally overriden by a user, so making a new package definition for a build flag that can be optional is an antipattern.

**Comment by ShalokShalom** [comment]
Oh, I see. Yeah, then there is definitely documentation missing.

**Comment by omar391** [comment]
I am new to Nixos. So maybe a naive question: 

how about we have some sort of profile to get a clear Linux repo to download the package/source including compiler configs instead of the default Nix repo? That would save us the hassle of configuring individual pkg to achieve the best performance for Linux hardware.

**Comment by blitz** [comment]
While I like a well optimized system like the next person, I personally would like to see some easy to run benchmarks for NixOS. Ideally, something we can plot over time. Otherwise, we will just cargo cult compiler flags...

Is there already a NixOS performance chat somewhere where some ideas can be discussed? Maybe that's the first step to get going. This is also related to #202526.
